File: packages/prebid_mobile_sdk_admob/ios/prebid_mobile_sdk_admob/Sources/prebid_mobile_sdk_admob/AdMobInterstitialManager.swift
import Flutter
import UIKit
import GoogleMobileAds
import PrebidMobile
import PrebidMobileAdMobAdapters

/// Handles AdMob-mediated interstitials over the
/// `prebid_mobile_sdk_admob/interstitial` method channel. Each ad is keyed by an
/// `adId` allocated on the Dart side; native events are pushed back over the
/// same channel.
class AdMobInterstitialManager: NSObject, FullScreenContentDelegate {

    private let channel: FlutterMethodChannel

    private var adUnits: [Int: MediationInterstitialAdUnit] = [:]
    private var mediationDelegates: [Int: AdMobMediationInterstitialUtils] = [:]
    private var interstitials: [Int: GoogleMobileAds.InterstitialAd] = [:]
    private var adIdByInterstitial: [ObjectIdentifier: Int] = [:]

    init(messenger: FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(
            name: "prebid_mobile_sdk_admob/interstitial",
            binaryMessenger: messenger
        )
        super.init()
        channel.setMethodCallHandler { [weak self] call, result in
            self?.handle(call, result)
        }
    }

    private func handle(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]
        let adId = args?["adId"] as? Int

        switch call.method {
        case "load":
            guard let adId = adId else {
                result(FlutterError(code: "no_ad_id", message: "Missing adId", details: nil))
                return
            }
            let configId = args?["configId"] as? String ?? ""
            let adMobAdUnitId = args?["adMobAdUnitId"] as? String ?? ""
            let isVideo = args?["isVideo"] as? Bool ?? false

            // 1. GMA request + Prebid mediation utils + ad unit.
            let gadRequest = Request()
            let mediationDelegate = AdMobMediationInterstitialUtils(gadRequest: gadRequest)
            let adUnit = MediationInterstitialAdUnit(
                configId: configId,
                mediationDelegate: mediationDelegate
            )
            adUnit.adFormats = isVideo ? [.video] : [.banner]

            adUnits[adId] = adUnit
            mediationDelegates[adId] = mediationDelegate

            // 2. Fetch demand, then load the AdMob interstitial.
            adUnit.fetchDemand { [weak self] _ in
                GoogleMobileAds.InterstitialAd.load(
                    with: adMobAdUnitId,
                    request: gadRequest
                ) { [weak self] ad, error in
                    guard let self = self else { return }
                    if let error = error {
                        self.send(adId, "onAdFailed", error: error.localizedDescription)
                        return
                    }
                    guard let ad = ad else { return }
                    ad.fullScreenContentDelegate = self
                    self.interstitials[adId] = ad
                    self.adIdByInterstitial[ObjectIdentifier(ad)] = adId
                    self.send(adId, "onAdLoaded")
                }
            }
            result(nil)

        case "show":
            if let adId = adId,
               let ad = interstitials[adId],
               let root = UIApplication.shared.keyWindow?.rootViewController {
                ad.present(from: root)
            }
            result(nil)

        case "destroy":
            if let adId = adId {
                if let ad = interstitials.removeValue(forKey: adId) {
                    adIdByInterstitial.removeValue(forKey: ObjectIdentifier(ad))
                }
                adUnits.removeValue(forKey: adId)
                mediationDelegates.removeValue(forKey: adId)
            }
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func send(_ adId: Int, _ event: String, error: String? = nil) {
        var payload: [String: Any] = ["adId": adId]
        if let error = error {
            payload["error"] = error
        }
        channel.invokeMethod(event, arguments: payload)
    }

    // MARK: - FullScreenContentDelegate

    func adWillPresentFullScreenContent(_ ad: FullScreenPresentingAd) {
        if let adId = adIdByInterstitial[ObjectIdentifier(ad as AnyObject)] {
            send(adId, "onAdDisplayed")
        }
    }

    func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
        if let adId = adIdByInterstitial[ObjectIdentifier(ad as AnyObject)] {
            send(adId, "onAdClosed")
        }
    }

    func adDidRecordClick(_ ad: FullScreenPresentingAd) {
        if let adId = adIdByInterstitial[ObjectIdentifier(ad as AnyObject)] {
            send(adId, "onAdClicked")
        }
    }

    func ad(
        _ ad: FullScreenPresentingAd,
        didFailToPresentFullScreenContentWithError error: Error
    ) {
        if let adId = adIdByInterstitial[ObjectIdentifier(ad as AnyObject)] {
            send(adId, "onAdFailed", error: error.localizedDescription)
        }
    }
}

File: packages/prebid_mobile_sdk_admob/android/src/main/kotlin/com/prebid/prebid_mobile_sdk_admob/AdMobInterstitialManager.kt
package com.prebid.prebid_mobile_sdk_admob

import android.app.Activity
import android.os.Bundle
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.FullScreenContentCallback
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.prebid.mobile.admob.AdMobMediationInterstitialUtils
import org.prebid.mobile.admob.PrebidInterstitialAdapter
import org.prebid.mobile.api.data.AdUnitFormat
import org.prebid.mobile.api.mediation.MediationInterstitialAdUnit
import java.util.EnumSet

/// Handles AdMob-mediated interstitials over the
/// `prebid_mobile_sdk_admob/interstitial` method channel. Each ad is keyed by an
/// `adId` allocated on the Dart side; native events are pushed back over the
/// same channel.
class AdMobInterstitialManager(
    messenger: BinaryMessenger,
    private val activityProvider: () -> Activity?,
) : MethodChannel.MethodCallHandler {

    private val channel = MethodChannel(messenger, "prebid_mobile_sdk_admob/interstitial")

    private class Holder(val adUnit: MediationInterstitialAdUnit) {
        var interstitial: InterstitialAd? = null
    }

    private val ads = mutableMapOf<Long, Holder>()

    init {
        channel.setMethodCallHandler(this)
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
        ads.values.forEach { it.adUnit.destroy() }
        ads.clear()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as? Map<*, *>
        val adId = (args?.get("adId") as? Number)?.toLong()

        when (call.method) {
            "load" -> {
                val activity = activityProvider()
                if (activity == null) {
                    result.error("no_activity", "No attached Activity to load the interstitial", null)
                    return
                }
                if (adId == null) {
                    result.error("no_ad_id", "Missing adId", null)
                    return
                }
                val configId = args?.get("configId") as? String ?: ""
                val adMobAdUnitId = args?.get("adMobAdUnitId") as? String ?: ""
                val isVideo = args?.get("isVideo") as? Boolean ?: false

                val extras = Bundle()
                val request = AdRequest.Builder()
                    .addNetworkExtrasBundle(PrebidInterstitialAdapter::class.java, extras)
                    .build()

                val mediationUtils = AdMobMediationInterstitialUtils(extras)
                val format = if (isVideo) AdUnitFormat.VIDEO else AdUnitFormat.BANNER
                val adUnit = MediationInterstitialAdUnit(
                    activity,
                    configId,
                    EnumSet.of(format),
                    mediationUtils,
                )
                val holder = Holder(adUnit)
                ads[adId] = holder

                adUnit.fetchDemand {
                    InterstitialAd.load(
                        activity,
                        adMobAdUnitId,
                        request,
                        object : InterstitialAdLoadCallback() {
                            override fun onAdLoaded(ad: InterstitialAd) {
                                holder.interstitial = ad
                                ad.fullScreenContentCallback = fullScreenCallback(adId)
                                send(adId, "onAdLoaded")
                            }

                            override fun onAdFailedToLoad(error: LoadAdError) {
                                holder.interstitial = null
                                send(adId, "onAdFailed", error.message)
                            }
                        },
                    )
                }
                result.success(null)
            }

            "show" -> {
                val activity = activityProvider()
                val ad = adId?.let { ads[it]?.interstitial }
                if (ad != null && activity != null) {
                    ad.show(activity)
                }
                result.success(null)
            }

            "destroy" -> {
                adId?.let { ads.remove(it)?.adUnit?.destroy() }
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun fullScreenCallback(adId: Long) = object : FullScreenContentCallback() {
        override fun onAdShowedFullScreenContent() = send(adId, "onAdDisplayed")
        override fun onAdDismissedFullScreenContent() = send(adId, "onAdClosed")
        override fun onAdClicked() = send(adId, "onAdClicked")
        override fun onAdFailedToShowFullScreenContent(error: AdError) =
            send(adId, "onAdFailed", error.message)
    }

    private fun send(adId: Long, event: String, error: String? = null) {
        val payload = mutableMapOf<String, Any?>("adId" to adId)
        if (error != null) payload["error"] = error
        channel.invokeMethod(event, payload)
    }
}
